lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -264,28 +268,38 @@ evaluateOrSkip(ConstExprStepEvaluator &stepEval,
 ///     instruction itself.
 static bool isInstructionFoldable(SingleValueInstruction *inst) {
   ASTContext &astContext = inst->getFunction()->getASTContext();
-  ;
   SILType silType = inst->getType();
 
   return (!silType.isAddress() && !isa<LiteralInst>(inst) &&
           !isa<StructInst>(inst) && !getStringMakeUTF8Init(inst) &&
           isIntegerOrStringType(silType, astContext));
 }
 
-/// Constant evaluate the instructions in the range 'first' to 'last'.
-/// Add foldable, constant-valued instructions discovered during the evaluation
-/// to the 'foldState' passed in.
+/// Constant evaluate the instructions from 'start' to the end of the function
+/// or until oslogMessage is released. Add foldable, constant-valued
+/// instructions discovered during the evaluation to the 'foldState' passed in.
 /// \returns error information for emitting diagnostics if the evaluation
 /// failed.
-static Optional<SymbolicValue> collectConstants(SILBasicBlock::iterator first,
-                                                SILBasicBlock::iterator last,
-                                                FoldState &foldState) {
+static Optional<SymbolicValue>
+collectConstants(SILInstruction *start, SingleValueInstruction *oslogMessage,
+                 FoldState &foldState) {
 
   ConstExprStepEvaluator &constantEvaluator = foldState.constantEvaluator;
+  SILBasicBlock::iterator currI = start->getIterator();
 
-  for (auto currI = first; currI != last;) {
+  // The loop will return when it sees a return instruction or a release value
+  // of oslogMessage.
+  while (true) {
     SILInstruction *currInst = &(*currI);
 
+    // Break if currInst is a release_value of 'oslogMessage'.
+    if (auto *releaseValueInst = dyn_cast<ReleaseValueInst>(currInst)) {
```

Comment:
1. For this to work when ownership is enabled, you need to at least handle destroy_value and probably need to look through copy_value.
2. Is this a struct or a class? If it is a class, you probably also need strong_release.

Reply:
> So to be clear: you marked the initializer with a semantic call so you could just constant fold it instead of inlining it?

Yes, that's right. One thing I would add is that the initializer was always constant evaluated but was mandatorily inlined earlier and then evaluated. Now it is marked with @_semantics and not mandatorily inlined (it can still be constant evaluated), and more importantly it is used to identify the range of instructions to evaluate.

Reply:
> * For this to work when ownership is enabled, you need to at least handle destroy_value and probably need to look through copy_value.
> * Is this a struct or a class? If it is a class, you probably also need strong_release.

This is a struct. I see. Thanks for the update, I will watch out for that.  Actually, this condition is more of a safe guard and in reality  the evaluation should break even before this instruction  as there will instructions that it cannot constant evaluate (due to the implementation of os_log) and will bail out earlier.